handlers/jiochat/jiochat.go

```diff
@@ -0,0 +1,323 @@
+		msg.WithAttachment(mediaURL)
+	}
+
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+}
+
+func resolveMediaID(mediaID string) string {
+	mediaURL, _ := url.Parse(fmt.Sprintf("%s/%s", jiochatURL, "media/download.action"))
+	mediaURL.RawQuery = url.Values{"media_id": []string{mediaID}}.Encode()
+	return mediaURL.String()
+}
+
+type jiochatAccess interface {
+	fetchAccessToken(*handler, courier.Channel) error
+	getAccessToken(*handler, courier.Channel) string
+}
+
+type jiochatClient struct{}
+
+type refreshTokenData struct {
+	GrantType    string `json:"grant_type"`
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+func (jCli *jiochatClient) fetchAccessToken(h *handler, channel courier.Channel) error {
+	time.Sleep(jiochatFetchTimeout)
+
+	tokenURL, _ := url.Parse(fmt.Sprintf("%s/%s", jiochatURL, "auth/token.action"))
+
+	refreshTokenData := &refreshTokenData{
+		GrantType:    "client_credentials",
+		ClientID:     channel.StringConfigForKey(configJiochatAppID, ""),
+		ClientSecret: channel.StringConfigForKey(configJiochatAppSecret, ""),
+	}
+
+	jsonBody, err := json.Marshal(refreshTokenData)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, tokenURL.String(), bytes.NewReader(jsonBody))
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	rr, err := utils.MakeHTTPRequest(req)
+
+	accessToken, err := jsonparser.GetString([]byte(rr.Body), "access_token")
+	if err != nil {
+		return err
+	}
+
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+	cacheKey := fmt.Sprintf("jiochat_channel_access_token:%s", channel.UUID().String())
+
+	_, err = rc.Do("set", cacheKey, accessToken, 7200)
+	return err
+}
+
+func (jCli *jiochatClient) getAccessToken(h *handler, channel courier.Channel) string {
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+	cacheKey := fmt.Sprintf("jiochat_channel_access_token:%s", channel.UUID().String())
+	accessToken, _ := redis.String(rc.Do("GET", cacheKey))
+	return accessToken
+}
+
+func (h *handler) findAccessToken(channel courier.Channel) string {
+	return h.jiochatClient.getAccessToken(h, channel)
+}
+
+func (h *handler) doFetchAccessToken(channel courier.Channel) error {
+	return h.jiochatClient.fetchAccessToken(h, channel)
+}
+
+type mtMsg struct {
+	MsgType string `json:"msgtype"`
+	ToUser  string `json:"touser"`
+	Text    struct {
+		Content string `json:"content"`
+	} `json:"text"`
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStatus, error) {
+	accessToken := h.findAccessToken(msg.Channel())
+
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored)
+	parts := handlers.SplitMsg(handlers.GetTextAndAttachments(msg), maxMsgLength)
```

Comment:
No MT media support?

Reply:
That requires to first upload the file and get an id, we did not support that on python I think we make the current behavior work and add the new support in a different PR

Reply:
Ok, I'm fine with that being another PR.